include/psa/crypto_extra.h

```diff
@@ -647,6 +647,90 @@ mbedtls_ecp_group_id mbedtls_ecc_group_of_psa( psa_ecc_family_t curve,
+
+/** The maximum value for a key identifier that is built into the
+ * implementation.
+ *
+ * See #MBEDTLS_PSA_KEY_ID_BUILTIN_MIN for more information.
+ */
+#define MBEDTLS_PSA_KEY_ID_BUILTIN_MAX          ((psa_key_id_t)0x7fffefff)
+
+/** A slot number identifying a key in a driver.
+ *
+ * Values of this type are used to identify built-in keys.
+ */
+typedef uint64_t psa_drv_slot_number_t;
+
+/** Platform function to obtain the data of a built-in key.
+ *
+ * You must implement this function if #MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS is
+ * enabled. This function is typically provided as part of a platform's
+ * system image.
+ *
+ * In the implementation of this function, to identify which key is requested,
+ * call psa_get_key_id(\p attributes) to obtain the key identifier
+ * \c key_id. Mbed TLS guarantees that
+ * #MBEDTLS_SVC_KEY_ID_GET_KEY_ID(\p key_id) is in the range from
+ * #MBEDTLS_PSA_KEY_ID_BUILTIN_MIN to #MBEDTLS_PSA_KEY_ID_BUILTIN_MAX.
+ *
+ * In a multi-application configuration
+ * (\c MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER is defined),
+ * this function must check that #MBEDTLS_SVC_KEY_ID_GET_OWNER_ID(\p key_id)
+ * is allowed to use the given key.
+ *
+ * \param[in,out] attributes    On entry, this is #PSA_KEY_ATTRIBUTES_INIT or
+ *                              an equivalent value, except that the key
+ *                              identifier field is set.
+ *                              On successful return, this function must set
+ *                              the attributes of the key: lifetime, type,
+ *                              bit-size, usage policy.
+ * \param[out] p_key_buffer     On successful return, this function must
+ *                              place a pointer to a buffer allocated with
+ *                              mbedtls_calloc() in \c *p_key_buffer, and
+ *                              fill this buffer with the key data.
+ * \param[out] key_buffer_size  On successful return, this function must
+ *                              place the size of the buffer \c *p_key_buffer
+ *                              in bytes.
+ *
+ * \retval #PSA_SUCCESS
+ *         The requested key identifier designates a built-in key.
+ *         In a multi-application configuration, the requested owner
+ *         is allowed to access it.
+ * \retval #PSA_ERROR_DOES_NOT_EXIST
+ *         The requested key identifier is not a built-in key which is known
+ *         to this function. If a key exists in the key storage with this
+ *         identifier, the data from the storage will be used.
+ * \retval (any other error)
+ *         Any other error is propagated to the function that requested the key.
+ *         Common errors include:
+ *         - #PSA_ERROR_INSUFFICIENT_MEMORY: unable to allocate the key buffer.
+ *         - #PSA_ERROR_NOT_PERMITTED: the key exists but the requested owner
+ *           is not allowed to access it.
+ */
+psa_status_t mbedtls_psa_platform_get_builtin_key(
```

Comment:
I am not very enthousiastic  with this platform function that
1) translate the key identifier to a (location, slot number) pair
2) call the driver corresponding to location to get the description of the built-in key identified by the slot number and return it.

What about for the platform function to do just 1. Then the core can call the drivers directly to get the key attributes and context.

The signature would be something like:
```c
psa_status_t mbedtls_psa_platform_get_builtin_key(
    mbedtls_svc_key_id_t id,
    psa_key_location_t *location,
    psa_drv_slot_number_t *slot_number);
```

Reply:
How does this work when deploying the same code image to multiple device models, some of which have a secure element and some not, so it isn't possible to tell at compile time whether the built-in key is in a secure element or not?

Reply:
Could you elaborate on the use case here to be sure we agree on what we are talking about?

Reply:
A device vendor wants to ship the same code image on two device models, one with a secure element and one without. (This is a real use case.) `mbedtls_psa_platform_get_builtin_key` then needs to have logic like this:
```
const root_key_slot_number[] = {42};
if driver.query_hardware_availability():
    attributes = driver.get_builtin_key_attributes(root_key_slot_number)
    key_material = &root_key_slot_number;
else:
    attributes = ...
    key_material = its_read(ROOT_KEY)
```

Reply:
ok I understand that. In the case without SE, which entity operates the key? I guess it could be the software implementation or any accelerator.

Reply:
The key is typically created during manufacturing using proprietary APIs, and the chip vendor writes the support code.

Reply:
Not sure to understand here, my question is what is the returned location for the key (returned as part of the key attributes) ?

Reply:
The location would be 0 (local storage).

Reply:
Arguably we could force “local” built-in keys (such as keys derived from a HUK) to use a separate location and be accessed via the driver interface. But that seems to me like it would impose a significant complexity overhead for a very common feature that doesn't need this kind of complexity.

Reply:
I think that is a good idea and this does not seem a significant complexity overhead to me.

The device vendor would have to provide two opaque drivers: one for the actual SE and one for the "storage only SE". In both cases, it has to write a get_builtin_key entry point thus the "storage onlyl" case doesn't bring anything new here and thinking about the "local" case as a "storage only SE" looks quite natural to me. 

Then the platform specific API in question here would contain only routing code thus would be simpler.

Another point is the handling of buffers. It seems that the current proposal relies on dynamic allocations while if the key attributes and data are retrieved by the core through driver calls we can probably also support the case without dynamic allocations along the lines of what is defined in the "Key format for opaque drivers".